src/DependencyInjection/DI/src/CallSiteFormatter.cs

```diff
@@ -0,0 +1,142 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Text;
+using Microsoft.Extensions.DependencyInjection.ServiceLookup;
+
+namespace Microsoft.Extensions.DependencyInjection
+{
+    internal class CallSiteFormatter: CallSiteVisitor<CallSiteFormatter.CallSiteFormatterContext, object>
+    {
+        internal static CallSiteFormatter Instance = new CallSiteFormatter();
+
+        private CallSiteFormatter()
+        {
+        }
+
+        public string Format(ServiceCallSite callSite)
+        {
+            var stringBuilder = new StringBuilder();
+
+            VisitCallSite(callSite, new CallSiteFormatterContext(stringBuilder, 0));
+
+            return stringBuilder.ToString();
+        }
+
+        protected override object VisitConstructor(ConstructorCallSite constructorCallSite, CallSiteFormatterContext argument)
+        {
+            argument.ContinueLine("new {0}", constructorCallSite.ImplementationType);
+
+            var newContext = argument.IncrementOffset();
+            foreach (var parameter in constructorCallSite.ParameterCallSites)
+            {
+                VisitCallSite(parameter, newContext);
+            }
+
+            return null;
+        }
+
+        protected override object VisitConstant(ConstantCallSite constantCallSite, CallSiteFormatterContext argument)
+        {
+            argument.ContinueLine("{0}", constantCallSite.DefaultValue ?? "<null>");
```

Comment:
Do we expect this value to be formattable? Most of the time these are user-provided instances of services, right? 

Perhaps just `<Constant: [TypeName]>` would be more appropriate.

Reply:
`object.ToString()` defaults to type name anyway.